### ISSUE TITLE
How to invoke native-image with --module argument via helidon-maven-plugin?

### ISSUE THREAD

**Issue by andrejusc** [oracle/helidon-build-tools#771]
For Helidon Maven plugin:
https://github.com/oracle/helidon-build-tools/tree/master/maven-plugins/helidon-maven-plugin
we have the goal `native-image`, which till recent was fine to use as input some jar from target folder.

But what if we want to utilize Java9+ modules and supply `--module` parameter instead of `-jar`? Looking at current Plugin's Mojo code - seems not yet covered:
https://github.com/oracle/helidon-build-tools/blob/master/maven-plugins/helidon-maven-plugin/src/main/java/io/helidon/build/maven/GraalNativeMojo.java

but possible if doing `native-image` binary invocation directly:
https://www.graalvm.org/22.1/reference-manual/native-image/Options/

and this part from it:

- native-image [options] --module `<module>[/<mainclass>]` [options] to build a native executable for a module. --module is equivalent to -m.

Any thoughts? At the moment in use `helidon-maven-plugin` v2.3.4.

**Comment by romain-grecourt** [comment]
Did you try using `additionalArgs` ?

**Comment by andrejusc** [comment]
I did, but that led to incorrect execution command like such:
```
Executing command: [/opt/graalvm-ee-java17-22.2.0/bin/native-image ... , --module=org.example.myfeature/org.example.MyFeature, -jar, /dev/target/service.jar, ...]
```
and such mix itself is incorrect from `native-image` invocation standpoint.

**Comment by andrejusc** [comment]
@romain-grecourt - if this issue could be prioritised - it will help to use both approaches for modularized and non-modularized native images building via same plugin.

**Comment by andrejusc** [comment]
Just to clarify - here we need to have 2 options now to be supported:
`--module-path` and `--module` for Java modularized jars to be used.

**Comment by andrejusc** [comment]
Do you have any update here? Maybe some Snapshot build I could try and report back?

**Comment by andrejusc** [comment]
If this could be somehow prioritized - it will help. Thank you for your support.

**Comment by andrejusc** [comment]
@romain-grecourt - with this closed - I cannot yet reach needed goal, i.e. have ability to specify arbitrary module path in a way like such:
```
          <configuration>
            <!-- Try here just for internal case first -->
            <execMode>module</execMode>
            <module>mymodule/mypackage.Main</module>
            <modulePath>myDir</modulePath>
          </configuration>
```
Could we re-open this and add that? See my clarified request 26 days ago here above.

**Comment by andrejusc** [comment]
To clarify my previous comment:

I.e. at the end I need this:
```
          <configuration>
            <!-- Try here just for internal case first -->
            <execMode>module</execMode>
            <module>mymodule/mypackage.Main</module>
            <modulePath>target/service.jar:target/libs</modulePath>
          </configuration>
```
to be passed like this to native-image generation:
```
--module-path target/service.jar:target/libs
```
and override project dependencies logic, which is currently introduced for module case.

**Comment by romain-grecourt** [comment]
@andrejusc You could have made that comment before we merged the pull requests...

@tvallin I can think of a few updates:
- `module` can be derived automatically, and the configuration can be used as an override, similar to `jarFile`
- `module` should probably imply `mainClass` (IIRC the last part of the command should be `module/mainClass`)
- add a new execution mode `none` that requires using `additionalArgs` ; the ordering with `-H:Name` may be an issue though

**Comment by andrejusc** [comment]
@romain-grecourt - I was awaiting some movement in this issue as it was stuck for some time and had a chance to build locally only today once PR was merged. I was assuming that need to have arbitrary `module-path` setting capability was obvious and that using instead project's dependencies not always could work within Maven multi-module projects and different phases of Maven execution.

Hoping that we could improve this further without another long wait time. Thank you for your support.

**Comment by romain-grecourt** [comment]
You don't have to wait for a PR to be merged in order to pull the changes.

---

FYI there is an official Maven plugin GraalVM Native Image: https://graalvm.github.io/native-build-tools/latest/maven-plugin.html ; You have the option of using the official plugin or use the exec-plugin and invoke `native-image` directly.

---

There is no direct equivalent to `java -jar` when using modules. When using `java -jar` the reference to `target/libs` is done in the `MANIFEST.MF` class-path entry, not in the plugin's defaults. Having the defaults for the new `module` `execMode` use `-m target/libs` would hard-code the reference to `target/libs`. This is not desired.

We implemented the default by following the same approach as the  official maven plugins (e.g. `maven-javadoc-plugin`). 

We will support your use-case through configuration:
- either by specifying `execMode=none` and supplying required arguments with `additionalArgs` 
- or by using a wildcard module exclusions and additional module-path entries

--

What specific problems are you having that are related to multi-module setup and/or Maven life-cycle ?

**Comment by andrejusc** [comment]
@romain-grecourt - my few comments to above:

>  You don't have to wait for a PR to be merged in order to pull the changes.

In normal Dev process I'm treating not merged PR as work in progress, which may or may not require further changes/corrections/etc. I'm also doing my best trying to evaluate 3+ different Open source projects changes together and so merged PRs in that case make my life easier. And as you folks have tons of other everyday activities.

> FYI there is an official Maven plugin GraalVM Native Image: https://graalvm.github.io/native-build-tools/latest/maven-plugin.html ; You have the option of using the official plugin or using the exec-plugin and invoke native-image directly.

I'm treating that Helidon as a project is Native image friendly thing and such doc in it's code base makes me think so, where it explicitly mentions presence/usage of `helidon-maven-plugin`: https://github.com/oracle/helidon/blob/3.0.0/docs/includes/guides/maven-build.adoc#the-helidon-maven-plugin 

You're right saying that other ways are possible to achieve native image generation, but then Helidon project's documentation should remove references to `helidon-maven-plugin` usage itself. And for my use case it will complicate things of gradual progression with what is already defined in related Maven pom files and try to change that to something else just because `helidon-maven-plugin` is having limitations. I prefer we work together on those limitations and eliminate them instead. In a hope that such could be of use for wider Dev community to utilize.

>  There is no direct equivalent to java -jar when using modules. 

True, but I'm not asking here about equivalent. I'm asking simply about some Mojo setting which pass what is in it into command line and into `--module-path` option of it. And I'm as consumer will be in full position to specify what is inside that setting.

> Having the defaults for the new module execMode use -m target/libs

In this case my `target/libs` sting was one of many examples, it not supposed to be hardcoded. Generic syntax of it would be `folder1_jar:folder2:folder3`, where `folder1_jar `could be `target/service.jar`, `folder2` could be `target/libs` and so on. Nothing default in that case, but configurable via my proposed `<modulePath>` setting of plugin configuration and in that case switching off currently implemented dependencies calculation.

> What specific problems are you having that are related to multi-module setup and/or Maven life-cycle ?

As you know - preparation of some JAR (in our case service.jar) happens usually within Maven `package` phase. Native image generation could also happen within same phase if related Maven `native-image` profile is active/effective. But in our case part of package phase does collection of needed dependencies into `target/libs` folder (and few more adjustments) and that is then fully ready for native image generation step to take. Those adjustments right now done to make sure none of dependencies are conflicting by any means from Java module angle - there are several such cases when they are.

**Comment by romain-grecourt** [comment]
> In normal Dev process I'm treating not merged PR as work in progress, which may or may not require further changes/corrections/etc. I'm also doing my best trying to evaluate 3+ different Open source projects changes together and so merged PRs in that case make my life easier. And as you folks have tons of other everyday activities.

A pull request is where we review the code. If you comment on a pull request asking about merging, that means you agree with the changes. Our everyday activities are irrelevant to this.

> Helidon project's documentation should remove references to `helidon-maven-plugin` usage itself. 

GraalVM Native Image 22 is not yet officially supported by Helidon, prior versions did not support `--module`. There is currently no reason to remove the usage of `helidon-maven-plugin:native-image`.

Note that this doesn't mean that won't switch to the official plugin in future versions of Helidon.

> Nothing default in that case, but configurable via my proposed `<modulePath>` setting of plugin configuration and in that case switching off currently implemented dependencies calculation.

You have a custom use-case, we do not want to support it as the default because it requires hard-coding a convention in the plugin. Like you said this will be done with additional configuration.

The configuration will be a bit different than what you've shared, though it should enable your use-case.

> I was assuming that need to have arbitrary module-path setting capability was obvious and that using instead project's dependencies not always could work within Maven multi-module projects and different phases of Maven execution.
>> What specific problems are you having that are related to multi-module setup and/or Maven life-cycle ?
>>> As you know - preparation of some JAR (in our case service.jar) happens usually within Maven `package` phase. Native image generation could also happen within same phase if related Maven `native-image` profile is active/effective. But in our case part of package phase does collection of needed dependencies into `target/libs` folder (and few more adjustments) and that is then fully ready for native image generation step to take. Those adjustments right now done to make sure none of dependencies are conflicting by any means from Java module angle - there are several such cases when they are.

The life-cycle that you describe is the one that is configured in Helidon applications (via parent pom). This is valid in a multi-module Maven project as well.

What are the adjustments that you mentioned ?

FYI the approach of using a mix of class-path / module-path computed from the project dependencies is used by official plugins such as `maven-compiler-plugin` and `maven-javadoc-plugin`. If you are currently having to work-around specific issues related to Java modules they may also surface with these plugins. E.g. try `javadoc:aggregate`

**Comment by andrejusc** [comment]
TL;DR ;)

> You have a custom use-case, we do not want to support it as the default because it requires hard-coding a convention in the plugin.

I don't see how my proposal of extra Mojo/configuration property `<modulePath>` leads to any hard-coding beside this config property name itself needs to be in code, which is usual thing for Mojos to be extended with. It's optional by nature for this use case. If I specify it - use it when invoking `native-image`. If not - use whatever logic you've implemented already.

> What are the adjustments that you mentioned ?

As you may know Helidon SE 2.5.x already (and soon Helidon MP 3.x) is based on `jakarta.*` namespace now versus `javax.*` namespace used by some libraries/jars before. If Maven via dependencies brings both namespaces in form of jars - that will potentially work for compilation purposes, but will fail for Java modules runtime bootstrap with so called Split package problem. So part of adjustment is to make sure only `jakarta.*` namespace related jars are in final POM dependencies and then inside `target/libs` folder (name of folder here arbitrary, could be anything) I have for my case. 

Other adjustment is modification of one of jars, which has "-native-" part in name as such is not valid from Java module angle and so I'm changing that on the fly by adding `Automatic-Module-Name` to it's Manifest. So then it's good Java modules citizen and both java/native-image are happy having it inside my target/libs area.

> FYI the approach of using a mix of class-path / module-path computed from the project dependencies is used by official plugins such as maven-compiler-plugin and maven-javadoc-plugin.

I simply want to be in position to override that if/when needed like previously we did with `jarFile` and `finalName` Mojo settings use cases.

**Comment by romain-grecourt** [comment]
> > We do not want to support it as the default because it requires hard-coding a convention in the plugin.
> 
> I don't see how my proposal of extra Mojo/configuration property `<modulePath>` leads to any hard-coding beside this config property name itself needs to be in code, which is usual thing for Mojos to be extended with.

You did not understand my statement. I'm talking about the default behavior, out of the box ; to justify why we implemented the default the way we did. 

The very next sentence:
> Like you said this will be done with additional configuration.

**Comment by romain-grecourt** [comment]
> > What are the adjustments that you mentioned ?
> 
> As you may know Helidon SE 2.5.x already (and soon Helidon MP 3.x) is based on `jakarta.*` namespace now versus `javax.*` namespace used by some libraries/jars before. So part of adjustment is to make sure only `jakarta.*` namespace related jars are in final POM dependencies and then inside `target/libs` folder (name of folder here arbitrary, could be anything) I have for my case.

Helidon does not mix `javax.*` and `jakarta.*` namespaces, 2.x supports `javax.*` 3.x supports `jakarta.*` ; the provided dependency management takes care of all the exclusions in order to provide a clean dependency tree (see `helidon-dependencies`).

If you are not using `helidon-dependencies` or an application parent pom, you have to do these exclusions yourself which can be tricky.

Assuming you have a clean dependency graph, I don't think this has any impact on your native-image build.

>> Other adjustment is modification of one of jars, which has "-native-" part in name as such is not valid from Java module angle and so I'm changing that on the fly by adding `Automatic-Module-Name` to it's Manifest. So then it's good Java modules citizen and both java/native-image are happy having it inside my target/libs area.

How are you modifying the jar name ? I'm assuming it's a dependency since you mention having it inside `target/libs`.

**Comment by andrejusc** [comment]
> Helidon does not mix javax.* and jakarta.* namespaces

Correct, but other projects/dependencies are not yet there. You could read more here about challenges: https://helidon.io/docs/v3/#/se/integrations/oci#oci-compatibility

> Assuming you have a clean dependency graph, I don't think this has any impact on your native-image build.

What I could guarantee (and that is enforced via our Maven package phase logic) is that dependencies collected into target/libs folder are passing such command:
```
java -p target/libs --validate-modules
```
which I cannot fully guarantee for project's Maven dependencies. Ideally it should be as well, but it's harder to prove without collecting them all into 1 single folder and using above command. Maven dependency plugin could help generate long line of involved dependencies, but I'm finding it's hard to use comparing to just listing/using 1 single folder.

> How are you modifying the jar name ?

Not jar name. MANIFEST.MF inside that jar via Maven jar plugin. This will be not needed with Helidon 3.x, but currently we're using Helidon 2.5.x. For more background - see this: https://github.com/oracle/helidon/issues/4425

>  I'm assuming it's a dependency since you mention having it inside target/libs

Yes, in some cases we have `helidon-graal-native-image-extension-2.5.X.jar` jar/dependency presence there, which is invalid right now from Java modules angle usage due to presence of Java key word "native" in it, cause it's automatic module right now. In Helidon 3.x it will be named odule, so problem will be gone.

More background on Java reserved keywords - here: https://www.w3schools.com/java/java_ref_keywords.asp

In short - any automatic module having in it's jar name such key word (for example "byte") will fail Java modules validation. That's why temporary workaround is possible (where applicable) via addition of extra Manifest attribute.

**Comment by romain-grecourt** [comment]
> > Helidon does not mix javax.* and jakarta.* namespaces
> 
> Correct, but other projects/dependencies are not yet there. You could read more here about challenges: https://helidon.io/docs/v3/#/se/integrations/oci#oci-compatibility

We are part of the Helidon dev team, you know that.. right ?

Like I said before, Helidon + managed dependencies are okay WRT to javax vs jakarta.
The oci-sdk is available in a shaded form that works around the actual issue.

> 
> > How are you modifying the jar name ?
> 
> Not jar name. MANIFEST.MF inside that jar via Maven jar plugin. This will be not needed with Helidon 3.x, but currently we're using Helidon 2.5.x. For more background - see this: [oracle/helidon#4425](https://github.com/oracle/helidon/issues/4425)

That issue has been fixed in 3.x and back ported in 2.x. There is a release available for both code lines.

Even if you are using an older version you should be able to fix for the entire build. I.e. you would repackage the offending jars, include module information and add dependency exclusions for the offending dependency and replace it with the repackaged one.

See also https://github.com/moditect/moditect#adding-module-descriptors-to-existing-jar-files

It seems to me that you have a few hacks that are not working properly with a Maven reactor or different life cycles (e.g. the jar is fixed only within a directory as part of a given execution). This is driving you down a path where you want to operate on a set of jars in a directory rather rely on the dependency graph.

**Comment by andrejusc** [comment]
> That issue has been fixed in 3.x and back ported in 2.x. There is a release available for both code lines.

You're right, I could see backport mentioned here:
https://github.com/oracle/helidon/releases/tag/2.5.2

but reality though for one of components I'm dealing with - it's still with 2.5.0. And update of it yet out of my control, so I'm dealing under constraints I have.
 
> Even if you are using an older version you should be able to fix for the entire build. I.e. you would repackage the offending jars...

Yes, it's kind of repackaging via invocation of `jar --update --manifest file ` command

> It seems to me that you have a few hacks that are not working properly with a Maven reactor or different life cycles

Who would be able to reach anything workable without them these days? ;) And in ideal world with Maven enforcer plugin/etc/etc there would be just 1 version of some dependency to deal with in clean dependency graph. Some (including myself) people understand importance of that, many others don't. It takes quite an effort to explain rationale and make it actually happen.

I appreciate all this thread/conversation and thank you for your support.

**Comment by romain-grecourt** [comment]
>> Even if you are using an older version you should be able to fix for the entire build. I.e. you would repackage the offending jars...

>Yes, it's kind of repackaging via invocation of `jar --update --manifest file ` command

It's about making a new Maven artifact that replaces the bad one. See https://github.com/oracle/helidon/tree/2.5.1/grpc/io.grpc

**Comment by tvallin** [comment]
See https://github.com/helidon-io/helidon-build-tools/pull/784

**Comment by andrejusc** [comment]
@tvallin - thank you for update. I'll try to test new code earliest this Wed, Sep 21. At the moment I'm trying to invoke `native-image` with modules directly via maven exec plugin (one of proposals in the middle of our long discussion by @romain-grecourt ) to see what else could be needed. One addition to original ask came in form to support `--features` argument. But I need to make it work initially via maven exec plugin before starting asking you/Romain on same.

Also - please see this my Graal related issue for `native-image` invocation, which could be related to all this activity: https://github.com/oracle/graal/issues/5004

**Comment by tvallin** [comment]
The documentation mention `--module <module>[/<mainclass>]` and not `--module=<module>[/<mainclass>]`. Have you tried it ?

**Comment by andrejusc** [comment]
@romain-grecourt - could you double-check if this went into v2.3.5 properly? It looks like tag 2.3.5 is pointing to some other commit: https://github.com/helidon-io/helidon-build-tools/tags ?

**Comment by romain-grecourt** [comment]
The fix for this issue is not included in any release yet.
2.3.5 was released 5 months ago (May 2022), the PR for this issue was merged 2 weeks ago.

**Comment by andrejusc** [comment]
I see, probably that was misunderstanding on my side as was consumed by other activities and only now returning to this. Do you want me to test this more or if you're confident about that PR - could you cut then next/2.3.6 release?

**Comment by andrejusc** [comment]
@romain-grecourt - I've tested now from 2.x master using such:
```
<execMode>none</execMode>
``` 
for my advanced use case and exact same arguments I had before for direct execution of `native-image` via maven exec plugin. All works fine. If you could cut 2.3.6 release to contain this enhancement - would be appreciated.

**Comment by romain-grecourt** [comment]
Done. The bits should be available in Maven Central shortly.

**Comment by andrejusc** [comment]
@romain-grecourt - now I've looked also into related Readme here for 2.3.6 tag:
https://github.com/helidon-io/helidon-build-tools/blob/2.3.6/helidon-maven-plugin/README.md?plain=1#L35
and so far it misses to mention `module` or `none` as available choices, having so far:
```
| execMode | enum | `jar` | Execution mode - `jar`, `jar-cp`, or `main` |
```
Not big deal for me - but could you cut some backlog ticket to adjust Readme later? Thank you

**Comment by tvallin** [comment]
See https://github.com/helidon-io/helidon-build-tools/issues/795

**Comment by romain-grecourt** [comment]
@andrejusc Can you please stop commenting on a closed issue. If you find anything else that is related, **please** file a new issue.
